### PR TITLE
Track notification counts per thread (implement MSC3773) (redo)

### PR DESCRIPTION
**Pull request by clokep**
Track notification counts on a per-thread basis, implementing MSC3773.

The overall design of this is to add a `thread_id` column to the `event_push_actions` (+ `event_push_actions_staging`) and `event_push_summary` tables. This allows the results to be segmented by the thread (by storing either the root event ID in the `thread_id` column or the special value `"main"`).

When retrieving counts of notifications we can then segment based on the thread, this is opt-in for the client by providing a sync flag. In the case the client doesn't want separate threads we simply sum across all threads + the main timeline.

The summarization code also needs to be updated to be per thread, instead of just per room.

Please see MSC3773 for a description of the API changes.

Part of #12550

This is essentially a rebase of #13181 on top of #13753 (see https://github.com/matrix-org/synapse/pull/13181#discussion_r963355433 for the rationale why it was separated).

We want this to land in Synapse v1.69.0 *the earliest* since it forces an updated schema which is landing in v1.68.0.

### CODE REVIEW

File: synapse/storage/databases/main/event_push_actions.py [hunk 1]
```diff
@@ -436,42 +454,70 @@ def _get_unread_counts_by_pos_txn(
...
 
-        summary_stream_ordering = 0
-        if row:
-            summary_stream_ordering = row[0]
-            counts.notify_count += row[1]
-            counts.unread_count += row[2]
+            # XXX All threads should have the same stream ordering?
+            max_summary_stream_ordering = max(
+                summary_stream_ordering, max_summary_stream_ordering
+            )
```
Comment (clokep) [inline synapse/storage/databases/main/event_push_actions.py:472]: There's a thread about this at https://github.com/matrix-org/synapse/pull/13181#discussion_r955192602, need to investigate.

File: synapse/storage/databases/main/event_push_actions.py [hunk 2]
```diff
@@ -436,42 +454,70 @@ def _get_unread_counts_by_pos_txn(
...
+        for summary_stream_ordering, notif_count, unread_count, thread_id in txn:
+            if thread_id == "main":
+                counts = NotifCounts(
+                    notify_count=notif_count, unread_count=unread_count
+                )
+            # TODO Delete zeroed out threads completely from the database.
+            elif notif_count or unread_count:
+                thread_counts[thread_id] = NotifCounts(
+                    notify_count=notif_count, unread_count=unread_count
+                )
```
Comment (clokep) [inline synapse/storage/databases/main/event_push_actions.py:467]: Copied from https://github.com/matrix-org/synapse/pull/13181#discussion_r955193227:

> Might be worth looking at this again briefly not that some other bugs have been fixed.

File: synapse/storage/databases/main/event_push_actions.py (hunk 1 above)
Comment (clokep) [reply synapse/storage/databases/main/event_push_actions.py:482]: @erikjohnston said:

> I think that stream ordering is the "max" stream ordering that we rotated from the event_push_actions table:
>
> https://github.com/matrix-org/synapse/pull/13181/files#diff-f121377d76a7b35c60092da2c4bf8c849544459a2c676cf8e87057aff24ece54R1218
>
> I think that means its equivalent to the stream ordering we rotated up to? They may differ but there should never be any relevant rows in EPA between the per-thread stream ordering and the max stream ordering?

To reword this a bit: for each `(user, room, thread)` tuple we store the latest stream ordering *in that thread* into the summary. So what @erikjohnston is saying is that this is pretty much equivalent to using the value stored in `event_push_summary_stream_ordering`, I think.

I'll try that out on `develop` (sans threading) first and see if that tracks.

Comment (clokep) [reply synapse/storage/databases/main/event_push_actions.py:482]: > this is pretty much equivalent to using the value stored in `event_push_summary_stream_ordering`, I think.

This isn't quite accurate though because receipts can be earlier than the latest summary (and some other edge-cases). Pretty much the `stream_ordering` is also used to let us know if the summary is valid.

Comment (clokep) [reply synapse/storage/databases/main/event_push_actions.py]: I clarified this comment in cb679e2379f981bc4fa3675b9ac37d642dc32985 (and removed the `XXX`). I think this code is valid, I just was not quite sure *why* it was valid. 👍

File: synapse/storage/databases/main/event_push_actions.py (hunk 2 above)
Comment (clokep) [reply synapse/storage/databases/main/event_push_actions.py]: It seems silly to delete rows that are likely to come back into play "soon", so I made two changes:

1. I updated the SQL query to ignore these rows (instead of returning them from the database and ignoring them in Python).
2. I added some code to delete empty push summaries that are over a month old. Those seem like rows that won't be needed again soon (and they can always be recreated).

**Comment by clokep** [comment]
@erikjohnston I set you as the reviewer since most of this PR is the same as #13181, which you had previously reviewed. You can push it back to the group if you'd like.

File: synapse/storage/databases/main/event_push_actions.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/event_push_actions.py]: > It seems silly to delete rows that are likely to come back into play "soon"

It's worth noting that an `UPDATE` is often more a less done via a `DELETE + INSERT` in the guts of postgres, so doing `UPDATE + UPDATE` is not obviously a win over `DELETE + INSERT`. Not sure if that changes anything?

Comment (erikjohnston) [reply synapse/storage/databases/main/event_push_actions.py]: Though actually HOT updates come into play here a bit, most likely: https://www.cybertec-postgresql.com/en/hot-updates-in-postgresql-for-better-performance/

File: synapse/storage/schema/main/delta/73/01thread_notifications_backfill.sql [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+-- Forces the background updates from 06thread_notifications.sql to run in the
+-- foreground as code will now require those to be "done".
+
+DELETE FROM background_updates WHERE update_name = 'event_push_backfill_thread_id';
+
+-- Overwrite any null thread_id columns.
+UPDATE event_push_actions_staging SET thread_id = 'main' WHERE thread_id IS NULL;
+UPDATE event_push_actions SET thread_id = 'main' WHERE thread_id IS NULL;
+UPDATE event_push_summary SET thread_id = 'main' WHERE thread_id IS NULL;
```
Comment (erikjohnston) [inline synapse/storage/schema/main/delta/73/01thread_notifications_backfill.sql:24]: Are these "fast" when the background update has been run?

File: synapse/storage/databases/main/event_push_actions.py [hunk 4]
```diff
@@ -463,42 +464,70 @@ def _get_unread_counts_by_pos_txn(
...
                 AND stream_ordering > ?
                 AND highlight = 1
+            GROUP BY thread_id
         """
         txn.execute(sql, (user_id, room_id, receipt_stream_ordering))
-        row = txn.fetchone()
-        if row:
-            counts.highlight_count += row[0]
+        for highlight_count, thread_id in txn:
+            if thread_id == "main":
```
Comment (erikjohnston) [inline synapse/storage/databases/main/event_push_actions.py:501]: We really should pull `"main"` out as a constant

File: synapse/storage/databases/main/event_push_actions.py [hunk 5]
```diff
@@ -463,42 +464,70 @@ def _get_unread_counts_by_pos_txn(
...
+                counts.highlight_count += highlight_count
+            elif highlight_count:
+                if thread_id in thread_counts:
```
Comment (erikjohnston) [inline synapse/storage/databases/main/event_push_actions.py]: Can we do this as:

```python
thread = thread_counts.get(thread_id)
if thread is not None:
    ...
```

File: synapse/storage/databases/main/event_push_actions.py [hunk 6]
```diff
@@ -463,42 +464,70 @@ def _get_unread_counts_by_pos_txn(
...
             txn, room_id, user_id, start_unread_stream_ordering
         )
 
-        counts.notify_count += notify_count
-        counts.unread_count += unread_count
+        for notif_count, unread_count, thread_id in unread_counts:
+            if thread_id == "main":
+                counts.notify_count += notif_count
+                counts.unread_count += unread_count
+            elif thread_id in thread_counts:
```
Comment (erikjohnston) [inline synapse/storage/databases/main/event_push_actions.py:525]: ditto

File: synapse/storage/databases/main/event_push_actions.py (hunk 2 above)
Comment (clokep) [reply synapse/storage/databases/main/event_push_actions.py]: > > It seems silly to delete rows that are likely to come back into play "soon"
> 
> It's worth noting that an `UPDATE` is often more a less done via a `DELETE + INSERT` in the guts of postgres, so doing `UPDATE + UPDATE` is not obviously a win over `DELETE + INSERT`. Not sure if that changes anything?

Maybe I'm over thinking, but I guess it lets the database engine potentially be smarter? 🤷

File: synapse/storage/schema/main/delta/73/01thread_notifications_backfill.sql (hunk 3 above)
Comment (clokep) [reply synapse/storage/schema/main/delta/73/01thread_notifications_backfill.sql:24]: They should be, I'm not really sure how to check that though?

File: synapse/storage/databases/main/event_push_actions.py [hunk 7]
```diff
@@ -1086,26 +1119,34 @@ def _handle_new_receipts_for_notifs_txn(self, txn: LoggingTransaction) -> bool:
...
-                    "unread_count": unread_count,
+                keyvalues={"user_id": user_id, "room_id": room_id},
+                updatevalues={
+                    "notif_count": 0,
+                    "unread_count": 0,
                     "stream_ordering": old_rotate_stream_ordering,
                     "last_receipt_stream_ordering": stream_ordering,
-                    "thread_id": "main",
                 },
             )
```
Comment (clokep) [inline synapse/storage/databases/main/event_push_actions.py:1137]: Hmm, actually I don't think this logic is actually correct. I think it needs to select on `stream_ordering` somehow?

Comment (clokep) [reply synapse/storage/databases/main/event_push_actions.py:1137]: From talking with @erikjohnston this is correct since the receipt applies to all threads in the room.

File: synapse/storage/schema/main/delta/73/01thread_notifications_backfill.sql (hunk 3 above)
Comment (clokep) [reply synapse/storage/schema/main/delta/73/01thread_notifications_backfill.sql:24]: Going to check this with a fake table to see how bad it is.

File: synapse/storage/schema/__init__.py [hunk 8]
```diff
@@ -87,9 +87,9 @@
 
 
 SCHEMA_COMPAT_VERSION = (
-    # The groups tables are no longer accessible, so synapses with SCHEMA_VERSION < 72
-    # could break.
-    72
+    # The threads_id column must exist for event_push_actions, event_push_summary,
+    # receipts_linearized, and receipts_graph.
+    73
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/schema/__init__.py:94]: We discussed this a bit and think it is OK.

Note that schema 73 was introduced in v1.68.0rc1.

File: synapse/storage/databases/main/event_push_actions.py (hunk 2 above)
Comment (clokep) [reply synapse/storage/databases/main/event_push_actions.py]: From discussion it seems that this is likely a bit of premature optimization, we're going to just file an issue for now.

File: synapse/storage/schema/main/delta/73/01thread_notifications_backfill.sql (hunk 3 above)
Comment (clokep) [reply synapse/storage/schema/main/delta/73/01thread_notifications_backfill.sql:24]: To test this I:

1. Took a large-ish database with an `event_push_actions` table of 340,977 rows.
2. Altered the table to add the nullable `thread_id` column.
3. Ran the above twice (with `EXPLAIN ANALYZE`).

Expectations are that the first time will take some time since it is filling in all the data, while the second time it is "fast" since there's nothing to do.

<details>
<summary>Results</summary>

```sql
postgres=# EXPLAIN ANALYZE UPDATE event_push_actions SET thread_id = 'main' WHERE thread_id IS NULL;
                                                           QUERY PLAN                                                            
---------------------------------------------------------------------------------------------------------------------------------
 Update on event_push_actions  (cost=0.00..9712.77 rows=1705 width=229) (actual time=21128.858..21128.859 rows=0 loops=1)
   ->  Seq Scan on event_push_actions  (cost=0.00..9712.77 rows=1705 width=229) (actual time=0.099..531.200 rows=340977 loops=1)
         Filter: (thread_id IS NULL)
 Planning Time: 0.154 ms
 Execution Time: 21128.894 ms
(5 rows)

Time: 21191.277 ms (00:21.191)
postgres=# EXPLAIN ANALYZE UPDATE event_push_actions SET thread_id = 'main' WHERE thread_id IS NULL;
                                                          QUERY PLAN                                                           
-------------------------------------------------------------------------------------------------------------------------------
 Update on event_push_actions  (cost=0.00..19966.42 rows=3505 width=229) (actual time=119.550..119.551 rows=0 loops=1)
   ->  Seq Scan on event_push_actions  (cost=0.00..19966.42 rows=3505 width=229) (actual time=119.548..119.549 rows=0 loops=1)
         Filter: (thread_id IS NULL)
         Rows Removed by Filter: 340977
 Planning Time: 0.073 ms
 Execution Time: 119.598 ms
(6 rows)

Time: 122.390 ms
```

</details>

In summary, the first run through took ~20 seconds (which isn't too terrible), but the second run through was ~0.1 seconds and seems to have done some smart filtering based on `thread_id IS NULL`.

File: synapse/storage/databases/main/event_push_actions.py (hunk 2 above)
Comment (clokep) [reply synapse/storage/databases/main/event_push_actions.py]: See #13875.

**Comment by clokep** [comment]
@erikjohnston I think this is ready for another review!

**Review by erikjohnston — APPROVED** [review]
Woop
